.github/bin/run-clang-tidy.cc

```diff
@@ -0,0 +1,226 @@
+  return result;
+}
+
+std::string GetContent(const fs::path &f) {
+  return GetContent(fopen(f.string().c_str(), "r"));
+}
+
+using hash_t = uint64_t;
+hash_t hash(const std::string &s) { return std::hash<std::string>()(s); }
+std::string toHex(uint64_t value, int show_lower_nibbles = 16) {
+  char out[16 + 1];
+  snprintf(out, sizeof(out), "%016" PRIx64, value);
+  return out + (16 - show_lower_nibbles);
+}
+
+std::optional<std::string> ReadAndVerifyTidyConfig(const fs::path &config) {
+  auto content = GetContent(config);
+  auto start_config = content.find("\nChecks:");
+  if (start_config == std::string::npos) {
+    std::cerr << "Not seen 'Checks:' in config " << config << "\n";
+    return std::nullopt;
+  }
+  if (content.find('#', start_config) != std::string::npos) {
+    std::cerr << "Comment found in check section of " << config << "\n";
+    return std::nullopt;
+  }
+  return content.substr(start_config);
+}
+
+fs::path GetCacheDir() {
+  if (const char *from_env = getenv("CACHE_DIR")) return fs::path(from_env);
+  if (const char *home = getenv("HOME")) {
+    if (auto cdir = fs::path(home) / ".cache/"; fs::exists(cdir)) return cdir;
+  }
+  return fs::path(getenv("TMPDIR") ?: "/tmp");
+}
+
+// Given a work-queue in/out-file, process it. Using system() for portability.
+void ClangTidyProcessFiles(const fs::path &content_dir, const std::string &cmd,
+                           std::list<filepath_contenthash_t> *work_queue) {
+  if (work_queue->empty()) return;
+  const int kJobs = std::thread::hardware_concurrency();
+  std::cerr << work_queue->size() << " files to process...";
+
+  std::mutex queue_access_lock;
+  auto clang_tidy_runner = [&]() {
+    for (;;) {
+      filepath_contenthash_t work;
+      {
+        const std::lock_guard<std::mutex> lock(queue_access_lock);
+        if (work_queue->empty()) return;
+        fprintf(stderr, "%5d\b\b\b\b\b", static_cast<int>(work_queue->size()));
+        work = work_queue->front();
+        work_queue->pop_front();
+      }
+      const fs::path output = content_dir / toHex(work.second);
+      const std::string tmp_out = output.string() + ".tmp";
+      std::string command = cmd + " " + work.first.string() + " > " + tmp_out;
+      command += " 2>/dev/null";
+      int r = system(command.c_str());
+#ifdef WIFSIGNALED
+      if (WIFSIGNALED(r) && (WTERMSIG(r) == SIGINT || WTERMSIG(r) == SIGQUIT)) {
+        break;  // got Ctrl-C
+      }
+#endif
+      std::error_code ec;
+      fs::rename(tmp_out, output, ec);
+    }
+  };
+  std::vector<std::thread> workers;
+  for (auto i = 0; i < kJobs; ++i) workers.emplace_back(clang_tidy_runner);
+  for (auto &t : workers) t.join();
+  fprintf(stderr, "     \n");  // Clean out progress counter.
+}
+
+int main(int argc, char *argv[]) {
+  std::error_code ignored_error;  // filesystem errors we don't care about.
+  if (!fs::exists("compile_commands.json")) {
+    std::cerr << "No compilation db found. First, run make-compilation-db.sh\n";
```

Comment:
`std::endl` is more canonical.

Reply:
I try to avoid as much as possible of the stream operator stuff as it mostly only adds clutter.
And in this case, it would break the line over 80 characters, so would occupy another line...